src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs

```diff
@@ -51,70 +50,171 @@ internal sealed class ServerMonitor : IServerMonitor
+            : this(
+                serverId,
+                endPoint,
+                connectionFactory,
+                serverMonitorSettings,
+                eventSubscriber,
+                roundTripTimeMonitor: new RoundTripTimeMonitor(
+                    connectionFactory,
+                    serverId,
+                    endPoint,
+                    Ensure.IsNotNull(serverMonitorSettings, nameof(serverMonitorSettings)).HeartbeatInterval,
+                    cancellationTokenSource.Token),
+                cancellationTokenSource)
+        {
+        }
+
+        public ServerMonitor(ServerId serverId, EndPoint endPoint, IConnectionFactory connectionFactory, ServerMonitorSettings serverMonitorSettings, IEventSubscriber eventSubscriber, IRoundTripTimeMonitor roundTripTimeMonitor, CancellationTokenSource cancellationTokenSource)
         {
+            _monitorCancellationTokenSource = cancellationTokenSource;
             _serverId = Ensure.IsNotNull(serverId, nameof(serverId));
             _endPoint = Ensure.IsNotNull(endPoint, nameof(endPoint));
             _connectionFactory = Ensure.IsNotNull(connectionFactory, nameof(connectionFactory));
             Ensure.IsNotNull(eventSubscriber, nameof(eventSubscriber));
+            _serverMonitorSettings = Ensure.IsNotNull(serverMonitorSettings, nameof(serverMonitorSettings));
+
+            _baseDescription = _currentDescription = new ServerDescription(_serverId, endPoint, reasonChanged: "InitialDescription", heartbeatInterval: serverMonitorSettings.HeartbeatInterval);
+            _roundTripTimeMonitor = Ensure.IsNotNull(roundTripTimeMonitor, nameof(roundTripTimeMonitor));
 
-            _baseDescription = _currentDescription = new ServerDescription(_serverId, endPoint, reasonChanged: "InitialDescription", heartbeatInterval: heartbeatInterval);
-            _heartbeatInterval = heartbeatInterval;
-            _timeout = timeout;
             _state = new InterlockedInt32(State.Initial);
             eventSubscriber.TryGetEventHandler(out _heartbeatStartedEventHandler);
             eventSubscriber.TryGetEventHandler(out _heartbeatSucceededEventHandler);
             eventSubscriber.TryGetEventHandler(out _heartbeatFailedEventHandler);
             eventSubscriber.TryGetEventHandler(out _sdamInformationEventHandler);
         }
 
-        /// <inheritdoc />
         public ServerDescription Description => Interlocked.CompareExchange(ref _currentDescription, null, null);
 
+        public object Lock => _lock;
+
+        // public methods
+        public void CancelCurrentCheck()
+        {
+            IConnection toDispose = null;
+            lock (_lock)
+            {
+                if (!Ensure.IsNotNull(_currentCheckCancellationTokenSource, nameof(_currentCheckCancellationTokenSource)).IsCancellationRequested)
```

Comment:
`Ensure` is only valid to use when checking arguments to a method (it throws some kind of `ArgumentException`).

This line should just be:

    if (!_currentCheckCancellationTokenSource.IsCancellationRequested)

Reply:
You still need to remove the call to `Ensure`

Reply:
sorry, I probably accidentally restored this line. Done